returnn/tf/util/basic.py

```diff
@@ -1339,6 +1340,142 @@ def __call__(self, shape, dtype=None, partition_info=None):
+  The default values are taken from RASR, see https://github.com/rwth-i6/rasr/blob/master/src/Signal/GammaTone.cc
+  :param float x:
+  :param float scaling_constant: A in [1]
+  :param float constant_of_integration: k in [1]
+  :param float slope: a in [1]
+  :return: frequency corresponding to given fractional length
+  :rtype: float
+  """
+  assert 0 <= x <= 1, "fractional length has to be between 0 and 1"
+  return scaling_constant * (np.power(10.0, slope * x) - constant_of_integration)
+
+
+def inv_greenwood_function(freq, scaling_constant=165.4, constant_of_integration=0.88, slope=2.1):
+  """
+  Inverse greenwood function, convert frequency to fractional length, see
+  Schlueter, Ralf, et al. "Gammatone features and feature combination for large vocabulary speech recognition."
+  ICASSP 2007 and also
+  https://en.wikipedia.org/wiki/Greenwood_function
+  The default values are taken from RASR, see https://github.com/rwth-i6/rasr/blob/master/src/Signal/GammaTone.cc
+  :param float freq:
+  :param float scaling_constant: A in [1]
+  :param float constant_of_integration: k in [1]
+  :param float slope: a in [1]
+  :return: fractional length corresponding to given frequency
+  :rtype: float
+  """
+  return np.log10(freq / scaling_constant + constant_of_integration) / slope
+
+
+class GammatoneFilterbankInitializer(init_ops.Initializer):
+  """
+  Initializer for a gammatone filterbank, e.g., to initialize weights of a convolutional layer.
+  Based on
+  [1] Schlueter, Ralf, et al. "Gammatone features and feature combination for large vocabulary speech recognition."
+  ICASSP 2007
+  """
+  def __init__(self, num_channels, length, sample_rate=16000, freq_max=7500., freq_min=100., normalization=True):
+    """
+    :param int num_channels: number of filters
+    :param int length: length of FIR filters in seconds
+    :param int sample_rate: sample rate of audio signal in Hz
+    :param float freq_max: maximum frequency of filterbank
+    :param float freq_min: minimum frequency of filterbank
+    :param bool normalization: normalize filterbanks to maximum frequency response of 0 dB
+    """
+    self.num_channels = num_channels
+    self.length = length
+    self.sample_rate = sample_rate
+    self.freq_max = freq_max
+    self.freq_min = freq_min
+    self.normalization = normalization
+
+  def __call__(self, shape, dtype=None, partition_info=None):
+    """
+    :param tuple[int] shape:
+    :param tf.DType dtype:
+    :param partition_info:
+    :rtype: tf.Tensor
+    """
+    center_freqs = self.center_frequencies(self.num_channels, self.freq_max, self.freq_min)
+    fbank = []
+    for freq in center_freqs:
+      fbank.append(self.gammatone_impulse_response(freq, self.length, self.sample_rate))
+    fbank = np.vstack(fbank)
+    if self.normalization:
+      fbank = self.normalize_filters(fbank)
+    fbank = np.reshape(np.transpose(fbank), shape).astype(dtype.as_numpy_dtype)
+    return tf.convert_to_tensor(fbank, dtype=dtype)
```

Comment:
It looks like most of the initializer is pure Numpy code, and the only relevant TF call happens in this line.

You should move everything to `util`. (Btw, maybe you can create a new file there, `util/numpy.py` or so. Or maybe `util/sig_proc.py`.)

And then, this `GammatoneFilterbankInitializer` in here, should be very simple (like most other initializers), and mostly call into the Numpy generator.